NDRange fixes

Oblomov opened the pull request:
This is a set of cleanups and fixes for `clEnqueueNDRangeKernel`. The main highlights are checks that for kernels with a required work-group size the user submitted that work-group size (which should fix #450), and a rewrite of the logic used to determine the work-group size in the NULL case, designed around the hard sizing constrains of dividing the global size, being less than the maximum allowed local size in each dimension while still trying to take the preferred work-group size multiple of the device into account.

franz commented:
Looks OK-ish to me, but i only stared at it for a while :)

It could be beneficial to move some of the guessing logic to a device driver hook, so each device can decide itself (e.g. HSA device + kernel with image arguments might be better of with "square" local sizes), but that's low on the TODO list.

I'm going to run this through some of pocl testsuites, if nothing breaks it's an ACK from me.

jrprice commented:
One concern I have with this patch is that (if I'm reading it correctly) it will greedily pick the largest work-group size it can when `local_size==NULL`. For example, if I enqueue an NDRange with global size `(4096,1,1)`, this will pick a work-group size of `(4096,1,1)`. This means you only have a single work-group, which means you will not be able to take advantage of any parallelism over multiple CPU cores.

Maybe this should also take the number of compute units into account?

Oblomov commented:
@jrprice I agree that the greedy algorithm might not always be the optimal one, and as @franz suggested there's plenty of room even for device-specific optimizations. Even on CPU, for example, there's no guarantee that splitting a 4096 global work size across multiple compute units would be better than on a single one, depending on amount of work, nature of memory accesses, potential effects of NUMA, etc.

But that's not the aim of this patchset. Here I focused on fixing what was definitely broken in the previous logic, rather than aiming at specific improvements. I actually strove to keep as close as possible to the original logic, both outside and inside the reduction loop. Further improvements are obviously possible, but a more careful analysis of the impact of the local size on execution is needed for those.

pjaaskel commented:
Nice to see someone working on this! @Oblomov can you first point out what was broken in the original logic? 

You are right that small grid sizes are problematic. We'd need to check how large kernel it is in terms of LLVM IR instructions (and if it has outer loops) for example to decide the number of cores to utilize with multiple WGs, but I think _always_ resorting to a single core has brings too major regression on performance with multicores.

Another aspect is that the original code (IIRC) tries to make the preferred width to multiple dimensions and now that I think about it, there's at least this logic behind: The vectorizer might be able to vectorize across some other dimension than X when/if it can utilize loop interchange if X is not suitable (e.g. due to memory access patterns not being sequential). I haven't checked how well loop interchange works in LLVM at the moment, but I've relied on it eventually taking care this case and also "outer loop vectorization".

Oblomov commented:
@pjaaskel The main problem of the original logic is that it did not take into account the two hard constraints (local work size must divide global work size, and local work size in each direction must not be larger than the maximum allowed size in that direction): instead, it started from a local size as large as the global work size and then halved (or clipped) directions until the overall local size fit within the hard third constraint (being no larger than the maximum overall local size).

This meant that the logic might produce invalid local sizes, e.g. by having a local size larger than the maximum allowed in that direction or a local size that did not divide the global size. This would be caught in the subsequent checks, but guessing an invalid local size is IMO wrong in principles.

So the main difference in my logic is that local size guessing now starts from the hard constraints (divisibility, and maximum size per direction). The logic used for the reduction until we fit in the third constraint is actually _mostly_ the same.

BTW, I feel this whole explanation should go in the commit message itself. Let me do just that.

pjaaskel commented:
OK, I see. Still I think there should be multiple WGs generated by default. I have a feeling the performance regression resulting from your patch will be more noticeable than the original issue has been.

Oblomov commented:
I don't think the new logic would introduce performance regressions, since for cases where the old logic produced valid work-group sizes, the new logic would produce workgroups with the same overall size: the main difference in the reduction is that we now prefer to halve Z first, then Y, then X instead of X first, then Y, then Z.

So, for example, assuming local size limits of (4096, 4096, 4096) with total limit of 4096, a global size of (4096, 2) would produce a local size of (2048, 2) with the old logic, and local size of (4096, 1) with the new logic. In terms of CU usage, there should be no difference.

Oblomov commented:
I've force-pushed an update for the message of the last commit, with a more detailed explanation of the differences between the old and new logic and highlighting some of the aspects discussed in the comments (particularly regarding the aim of the patch and the possibility of performance regression).

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -44,6 +44,38 @@
+}
+
+/* Find the largest divisor of K which is less than L */
+static inline size_t upper_divisor(size_t K, size_t L)
```
Why capital K and L?

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -110,15 +147,63 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
-        goto DETERMINE_LOCAL_SIZE;
+
+      POCL_RETURN_ERROR_ON((local_x * local_y * local_z > max_group_size),
+        CL_INVALID_WORK_GROUP_SIZE, "Local worksize dimensions exceed device's max workgroup size\n");
```
Check long lines.

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -110,15 +147,63 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
-  else
+
+  /* If the kernel has the reqd_work_group_size attribute, then the local
+   * work size _must_ be speficied, and it _must_ match the attribute specification
```
Typo speficied

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -110,15 +147,63 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+   * 2.0 support scenarios, so keep the conditional around, commented
+   */
+  else if (local_work_size == NULL)
+    /* || (local_x > global_x || local_y > global_y || local_z > global_z */
```
Please remove commented out code.

Oblomov commented:
Updated. Hopefully I addressed all comments. I've also included a preliminary “try to fill all CUs” algorithm after the greedy one. (It's intentionally separate, since the logic for the two is quite distinct.)

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -129,81 +217,212 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+
+          while (splits > 1)
+            {
+              TRY_SPLIT(z) else TRY_SPLIT(y) else TRY_SPLIT(z)
```
Should it be TRY_SPLIT (x) in the last else?

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -129,81 +217,212 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+      while (((splits = ncus/(nwg_x*nwg_y*nwg_z)) > 1) &&
+             (local_x*local_y*local_z > splits*min_group_size))
+        {
+          /* Very simple splitting approach: find a dimension divisible
```
I think this is OK, but a tad better would be to favour splits that maintain as many dimensions as possible the preferred SIMD width -- but this can be optimized forever.

pjaaskel commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -129,81 +217,212 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+      /* Pick a minimum work-group size of 4 times the preferred work-group size
+       * multiple, under the assumption that this would be a good candidate
+       * below which a Compute Unit will not do enough work. */
+      const size_t min_group_size = 4*preferred_wg_multiple;
```
Please check the GNU C style. There are some missing white spaces before and after operators and before (.

https://www.gnu.org/prep/standards/html_node/Formatting.html#Formatting

jrprice commented on lib/CL/clEnqueueNDRangeKernel.c:
```diff
@@ -129,81 +217,212 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+
+      cl_uint ncus; /* number of compute units */
+      prop_err = POname(clGetDeviceInfo)(command_queue->device,
+        CL_DEVICE_MAX_COMPUTE_UNITS, sizeof(cl_uint), &ncus, NULL);
```
Is there any reason not to just access `device->max_compute_units` directly?

pjaaskel commented:
Oh, and please describe the update in CHANGES.
